### PR TITLE
Functions

### LINKED ISSUE #5: "Function" type modules
It could conveniently be a subclass, which only overwrites the Verify function by checking that all reactions using inputs are catalysts

### PR DESCRIPTION
**Pull request by adramaxxx**
Better exception handling is needed, but i felt like this should be an issue of its own

### CODE REVIEW

**Comment by adramaxxx** [comment]
need to add one more test though, hang tight

File: src/parser.yy [hunk 1]
```diff
@@ -96,11 +97,18 @@ void InsertToSpecieMap(speciesRatios &ratio, SpeciesPair &toInsert) {
 
 modules  : module
          | modules module
+         | functions
          ;
 
 module : T_DMODULE "name" "{" properties "}" { drv.currentModule.name = $2; drv.FinishParsingModule(); }
 	   ;
 
+functions : function
```
Comment (RasmusRendal) [inline src/parser.yy]: Maybe you don't need to have the `functions` non-terminal at all in the grammar. After all, it only exists to implement left-recursion, and that is already implemented in `modules`.

File: src/module.cpp [hunk 2]
```diff
@@ -114,3 +114,12 @@ void Module::ApplyCompositions() {
 		compositions.pop_back();
 	}
 }
+
+void Module::VerifyFunction() {
+	for (const auto &input : inputSpecies) {
+		for (auto reaction : reactions) {
+			if (reaction.reactants[input] != reaction.products[input])
```
Comment (RasmusRendal) [inline src/module.cpp:121]: I'm worried about an `std::out_of_range` exception here. What happens if you have a function
```
function MultiplyTwo {
    input: a;
    private: b;
    output: c;
    concentrations: {
        b := 2;
    }
    reactions: {
        a + b -> a + b + c;
        c -> 0;
    }
}
```
Or any other function, which contains a reaction that has no instance of all of the input species.

**Review by RasmusRendal — CHANGES_REQUESTED** [review]
I like the way of adding a FinishParsingFunction method to the module. I was probably going to add a whole new subclass, which I now see is needlessly complex.

File: src/module.cpp (hunk 2 above)
Comment (adramaxxx) [reply src/module.cpp:121]: Works fine here. I can implement a test to combat this in the feature?

File: src/parser.yy (hunk 1 above)
Comment (adramaxxx) [reply src/parser.yy]: bison get's angry if it is non-terminal :S

File: src/module.cpp (hunk 2 above)
Comment (RasmusRendal) [reply src/module.cpp:121]: Huh, looking at the documentation for `std::map`, the [] operator does indeed insert a new member with the default value.

I'd say add the test anyway. Then if we switch away from maps or something, we won't have an unnoticed bug. Besides, more tests = more better;

File: src/parser.yy (hunk 1 above)
Comment (RasmusRendal) [reply src/parser.yy]: You can't just add the rule `T_DFUNCTION "name" "{" properties "}" { drv.currentModule.name = $2; drv.FinishParsingFunction(); }` to module?

**Review by RasmusRendal — CHANGES_REQUESTED** [review]

File: src/parser.yy (hunk 1 above)
Comment (adramaxxx) [reply src/parser.yy]: i had not thought of this, since i thought they had to be distinct from one another, but the token makes them distinct

**Review by RasmusRendal — APPROVED** [review]
Great stuff

**Pull request merged**